src/types/debug-utils.ts

```diff
@@ -0,0 +1,171 @@
+  txnGroupSources: AVMDebuggerSourceMapEntry[]
+
+  constructor(txnGroupSources: AVMDebuggerSourceMapEntry[]) {
+    this.txnGroupSources = txnGroupSources
+  }
+
+  static fromDict(data: AVMDebuggerSourceMapDict): AVMDebuggerSourceMap {
+    return new AVMDebuggerSourceMap(
+      data['txn-group-sources'].map((item) => new AVMDebuggerSourceMapEntry(item['sourcemap-location'], item['hash'])),
+    )
+  }
+
+  toDict(): AVMDebuggerSourceMapDict {
+    return { 'txn-group-sources': this.txnGroupSources.map((item) => JSON.parse(item.toString())) }
+  }
+}
+
+export class PersistSourceMapInput {
+  teal: string
+  appName: string
+  private _fileName: string
+
+  constructor(teal: string, appName: string, fileName: string) {
+    this.teal = teal
+    this.appName = appName
+    this._fileName = this.stripTealExtension(fileName)
+  }
+
+  get fileName(): string {
+    return this._fileName
+  }
+
+  set fileName(value: string) {
+    this._fileName = this.stripTealExtension(value)
+  }
+
+  private stripTealExtension(fileName: string): string {
+    return fileName.endsWith('.teal') ? fileName.replace('.teal', '') : fileName
+  }
+}
+
+export async function loadOrCreateSources(sourcesPath: string): Promise<AVMDebuggerSourceMap> {
+  try {
+    const data = JSON.parse(await fs.promises.readFile(sourcesPath, 'utf8'))
+    return AVMDebuggerSourceMap.fromDict(data)
+  } catch (error: unknown) {
+    const err = error as ErrnoException
+
+    if (err.code === 'ENOENT') {
+      return new AVMDebuggerSourceMap([])
+    } else {
+      throw error
+    }
+  }
+}
+
+export async function upsertDebugSourcemaps(sourcemaps: AVMDebuggerSourceMapEntry[], projectRoot: string): Promise<void> {
+  const sourcesPath = path.join(projectRoot, ALGOKIT_DIR, SOURCES_DIR, SOURCES_FILE)
+  const sources = await loadOrCreateSources(sourcesPath)
+
+  for (const sourcemap of sourcemaps) {
+    const sourceFilePath = path.resolve(sourcemap.location)
+    try {
+      await fs.promises.access(sourceFilePath)
+      const index = sources.txnGroupSources.findIndex((item) => item.equals(sourcemap))
+      if (index === -1) {
+        sources.txnGroupSources.push(sourcemap)
+      } else {
+        sources.txnGroupSources[index] = sourcemap
+      }
+    } catch (error: unknown) {
+      const err = error as ErrnoException
+
+      if (err.code === 'ENOENT') {
+        const index = sources.txnGroupSources.findIndex((item) => item.equals(sourcemap))
+        if (index !== -1) {
+          sources.txnGroupSources.splice(index, 1)
+        }
+      } else {
+        throw error
+      }
+    }
+  }
+
+  await fs.promises.writeFile(sourcesPath, JSON.stringify(sources.toDict()), 'utf8')
+}
+
+export async function writeToFile(filePath: string, content: string): Promise<void> {
+  await fs.promises.mkdir(path.dirname(filePath), { recursive: true })
+  await fs.promises.writeFile(filePath, content, 'utf8')
+}
+
+export async function buildAVMSourcemap({
```

Comment:
I don't think these exported functions really belong in the type directory.

Reply:
@neilcampbell whats a better spot to place them in this case? I assume at the root of `src`?